src/link-nodes.js

```diff
@@ -97,23 +91,26 @@ export const recordRemoveChild = (node, container) => {
 }
 
 /**
- * @param  {!Node} node
+ * @param  {!Node|DocumentFragment} node
+ * @param  {!Node|DocumentFragment=} root
  */
-export const recordChildNodes = (node) => {
+export const recordChildNodes = (node, root) => {
   const nodeData = ensureShadyDataForNode(node);
-  if (nodeData.firstChild === undefined) {
-    // remove caching of childNodes
-    nodeData.childNodes = null;
-    const first = nodeData.firstChild = node[utils.NATIVE_PREFIX + 'firstChild'] || null;
-    nodeData.lastChild = node[utils.NATIVE_PREFIX + 'lastChild'] || null;
-    patchInsideElementAccessors(node);
-    for (let n = first, previous; n; (n = n[utils.NATIVE_PREFIX + 'nextSibling'])) {
-      const sd = ensureShadyDataForNode(n);
-      sd.parentNode = node;
-      sd.nextSibling = n[utils.NATIVE_PREFIX + 'nextSibling'] || null;
-      sd.previousSibling = previous || null;
-      previous = n;
-      patchOutsideElementAccessors(n);
-    }
+  if (!root && nodeData.firstChild !== undefined) {
+    return;
+  }
+  root = root || node;
```

Comment:
Confusing to call this `root` here, it's really `parent`.  L110 could just be `sd.parentNode = root || node`.

It might be good to call the `root` argument `adoptedRoot` or something like that?

Reply:
Changed to `adoptedParent`